ouroboros-consensus/test-consensus/Test/Dynamic/Util.hs

```diff
@@ -241,6 +241,56 @@ incCrowdedRun :: SlotNo -> CrowdedRun -> CrowdedRun
 incCrowdedRun slot (CrowdedRun Nothing)          = CrowdedRun (Just (slot, 1))
 incCrowdedRun _    (CrowdedRun (Just (slot, n))) = CrowdedRun (Just (slot, n + 1))
 
+-- | The longest 'CrowdedRun' of a 'LeaderSchedule'. The tiebreaker favors
+-- earlier runs. The \"length\" of a run is the number of slots in it that have
+-- more than one leader. The run may also have slots with no leader in it, but
+-- these don't make it longer. A run has no slots with exactly one leader.
+--
+-- Such a run can drive a network away from consensus. If its long enough, it
+-- might even prevent consensus thereafter. The length of the run is defined as
+-- a number of slots, but it is also interpreted as an upper bound on the
+-- number of blocks that can differ among the competing chains' suffixes at the
+-- end of the run. Thus, if the length of a run exceeds the security parameter
+-- @k@, then that part of the leader schedule itself may be enough to prevent
+-- the network from ever re-establishing consensus. The method of failure is as
+-- follows.
+--
+-- * Assume the network is in consensus as of slot @s - 1@: all nodes have the
+-- same current chain.
+--
+-- * Let the slot @s@ have @N > 1@ leaders. Each leader forges a block and adds
+-- it to its own chain. ChainSync distributes these @N@ new blocks (headers,
+-- actually) to the non-leaders, each of which adopts one of them (the first to
+-- arrive?) as the new tip of its chain, since it makes it longer. The @N@
+-- leaders have created @N@ competing current chains amongst the network. ("Up
+-- to @N@ ..."; could a leader receive and adopt another leader's block before
```

Comment:
> ("Up to `N` ..."; could a leader receive and adopt another leader's block before it adopts the one it forged?)

That is possible indeed (at least theoretically). We take this into account when producing a block: https://github.com/input-output-hk/ouroboros-network/blob/4e5b99098cc4e4bbf400a9a8e98c42dba6ba574e/ouroboros-consensus/src/Ouroboros/Consensus/Node.hs#L365

So I would include your comment in parentheses.